src/discovery.rs

```diff
@@ -0,0 +1,75 @@
+use crate::errors::KaboodleError;
+use crate::networking::create_broadcast_sockets;
+use crate::structs::{SwimBroadcast, SwimEnvelope};
+use bytes::Bytes;
+use if_addrs::Interface;
+use tokio::time::timeout;
+
+use std::{
+    net::SocketAddr,
+    time::{Duration, Instant},
+};
+use tokio::net::UdpSocket;
+
+/// How large a buffer to use when reading from our sockets; messages larger than this will be
+/// truncated.
+const INCOMING_BUFFER_SIZE: usize = 1024;
+
+/// How often to re-broadcast our Probe message if we don't get a response.
+const REBROADCAST_INTERVAL: Duration = Duration::from_millis(10000);
+
+/// Discovers one member of the mesh on the given port and interface, without actually joining
+/// the mesh. Use this if you simply need to discover a member of the mesh but do not want to
+/// join the mesh yourself.
+pub async fn discover_mesh_member(
+    broadcast_port: u16,
+    interface: Interface,
+) -> Result<(SocketAddr, Bytes), KaboodleError> {
+    let sock = {
+        let ip = interface.ip();
+        UdpSocket::bind(format!("{ip}:0")).await?
+    };
+    let self_addr = sock
+        .local_addr()
+        .expect("Failed to get local socket address");
+    let (_broadcast_in_sock, broadcast_out_sock, broadcast_addr) =
+        create_broadcast_sockets(&interface, &broadcast_port)?;
+
+    let mut buf = [0; INCOMING_BUFFER_SIZE];
+    let mut last_broadcast_time = Instant::now() - REBROADCAST_INTERVAL - Duration::from_secs(1);
+    let out_bytes =
+        bincode::serialize(&SwimBroadcast::Probe(self_addr)).expect("Failed to serialize");
+
+    loop {
```

Comment:
Random idea, might be flawed: Even before starting the loop, this function could check if `localhost` or `::1` are responsive... significantly speeding things up (skipping the need for discovery entirely) if there is a local `serval-agent` running!